source/MaterialXView/Handlers/ViewHandler.cpp

```diff
@@ -193,75 +144,72 @@ bool ViewHandler::invertGeneralMatrix(const Matrix4x4& m, Matrix4x4& im) const
         {
             // Singular
-            makeIdentityMatrix(im);
+            im = Matrix44::IDENTITY;
             return false;
         }
 
         for (unsigned int j = 0; j < 3; j++)
         {
-            t[i*R_OFF+j] /= f;
-            im[i*R_OFF+j] /= f;
+            t[i][j] /= f;
+            im[i][j] /= f;
         }
 
         for (unsigned int j = 0; j < i; j++)
         {
-            f = t[j*R_OFF+i];
+            f = t[j][i];
 
             for (unsigned int k = 0; k < 3; k++)
             {
-                t[j*R_OFF+k] -= f * t[i*R_OFF+k];
-                im[j*R_OFF+k] -= f * im[i*R_OFF+k];
+                t[j][k] -= f * t[i][k];
+                im[j][k] -= f * im[i][k];
             }
         }
     }
 
     return true;
 }
 
-bool ViewHandler::invertMatrix(const Matrix4x4& m, Matrix4x4& im) const
+bool ViewHandler::invertMatrix(const Matrix44& m, Matrix44& im) const
 {
-    // Row offset
-    const unsigned int R_OFF = 4;
-
     // Not-affine, use general computation
-    if (m[0 * 4 + 3] != 0.0f || m[1 * 4 + 3] != 0.0f || m[2 * 4 + 3] != 0.0f || m[3 * 4 + 3] != 1.0f)
+    if (m[0][3] != 0.0f || m[1][3] != 0.0f || m[2][3] != 0.0f || m[3][3] != 1.0f)
     {
         return invertGeneralMatrix(m, im);
     }
 
     // Affine matrix inverse
-    im[0] = m[1*R_OFF+1] * m[2*R_OFF+2] - m[2*R_OFF+1] * m[1*R_OFF+2];
-    im[0] = m[2*R_OFF+1] * m[2] - m[1] * m[2*R_OFF+2];
-    im[0] = m[1] * m[1*R_OFF+2] - m[1*R_OFF+1] * m[2];
-    im[0] = 0.0f;
+    im[0][0] = m[1][1] * m[2][2] - m[2][1] * m[1][2];
+    im[0][1] = m[2][1] * m[0][2] - m[0][1] * m[2][2];
+    im[0][2] = m[0][1] * m[1][2] - m[1][1] * m[0][2];
+    im[0][3] = 0.0f;
 
-    im[0] = m[2*R_OFF] * m[1*R_OFF+2] - m[1*R_OFF] * m[2*R_OFF+2];
-    im[0] = m[0] * m[2*R_OFF+2] - m[2*R_OFF] * m[2];
-    im[0] = m[1*R_OFF] * m[2] - m[0] * m[1*R_OFF+2];
-    im[0] = 0.0f;
+    im[1][0] = m[2][0] * m[1][2] - m[1][0] * m[2][2];
+    im[1][1] = m[0][0] * m[2][2] - m[2][0] * m[0][2];
+    im[1][2] = m[1][0] * m[0][2] - m[0][0] * m[1][2];
+    im[1][3] = 0.0f;
 
-    im[0] = m[1*R_OFF] * m[2*R_OFF+1] - m[2*R_OFF] * m[1*R_OFF+1];
-    im[0] = m[2*R_OFF] * m[1] - m[0] * m[2*R_OFF+1];
-    im[0] = m[0] * m[1*R_OFF+1] - m[1*R_OFF] * m[1];
-    im[0] = 0.0f;
+    im[2][0] = m[1][0] * m[2][1] - m[2][0] * m[1][1];
+    im[2][1] = m[2][0] * m[0][1] - m[0][0] * m[2][1];
+    im[2][2] = m[0][0] * m[1][1] - m[1][0] * m[0][1];
+    im[2][3] = 0.0f;
 
-    im[0] = 0.0f;
-    im[0] = 0.0f;
-    im[0] = 0.0f;
-    im[0] = 1.0f;
+    im[3][0] = 0.0f;
+    im[3][1] = 0.0f;
+    im[3][2] = 0.0f;
+    im[3][3] = 1.0f;
 
-    float r = m[0] * im[0] + m[1] * im[1*R_OFF] + m[2] * im[2*R_OFF];
+    float r = m[0][0] * im[0][0] + m[0][1] * im[0][2] * im[2][0];
 
     if (std::abs(r) >= 1.0f)
     {
         for (unsigned int i = 0; i < 3; i++)
         {
             for (unsigned int j = 0; j < 3; j++)
             {
-                im[i*R_OFF+j] /= r;
+                im[i][j] /= r;
```

Comment:
Can we be sure the compiler will find and optimize this? Converting 9 divisions to a one time division and 9 multiply. We probably can but just checking what you think.

Reply:
My initial thought is that the compiler wouldn't be allowed to perform this optimization, because the invert-and-multiply would generate different floating-point precision errors than the requested divide.

Reply:
Good point. I assume that difference in precision is not something we need to care about here. So in this case we could move the division out of the loop manually then.